server/model/model.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"fmt"
+
+	cjson "github.com/gibson042/canonicaljson-go"
+	"github.com/juju/errors"
+	"github.com/pingcap-incubator/tiup/pkg/log"
+	"github.com/pingcap-incubator/tiup/pkg/repository/v1manifest"
+	"github.com/pingcap-incubator/tiup/pkg/utils"
+	"github.com/pingcap-incubator/tiup/server/store"
+)
+
+// Model defines operations on the manifests
+type Model interface {
+	UpdateComponentManifest(component string, manifest *ComponentManifest) error
+	UpdateRootManifest(manifest *RootManifest) error
+	UpdateIndexManifest(func(*IndexManifest) *IndexManifest) error
+	UpdateSnapshotManifest(func(*SnapshotManifest) *SnapshotManifest) error
+	UpdateTimestampManifest() error
+}
+
+type model struct {
+	txn  store.FsTxn
+	keys map[string]*v1manifest.KeyInfo
+}
+
+// New returns a object implemented Model
+func New(txn store.FsTxn, keys map[string]*v1manifest.KeyInfo) Model {
+	return &model{txn, keys}
+}
+
+func (m *model) UpdateComponentManifest(component string, manifest *ComponentManifest) error {
+	snap, err := m.ReadSnapshotManifest()
+	if err != nil {
+		return err
+	}
+	lastVersion := snap.Signed.Meta[fmt.Sprintf("/%s.json", component)].Version
+	if manifest.Signed.Version != lastVersion+1 {
+		log.Debugf("Component version not expected, expect %d, got %d", lastVersion+1, manifest.Signed.Version)
+		return ErrorConflict
+	}
+	return m.txn.WriteManifest(fmt.Sprintf("%d.%s.json", manifest.Signed.Version, component), manifest)
```

Comment:
There is a writer function in `v1manifest` that handles version number automatically

Reply:
In this function, we do some extra things: record the filename we modified to avoid update conflict